cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,336 @@
+
+        # Perform the process until there are no active extant lineages left
+        while len(current_lineages) > 0:
+            # If number of extant lineages is the stopping criterion, at the
+            # first instance of having n extant tips, stop the experiment
+            # and set the total lineage time for each lineage to be equal to
+            # the minimum, to produce ultrametric trees. Also, the birth_scale
+            # parameter of each leaf is rolled back to equal its parent's.
+            if num_extant:
+                if len(current_lineages) == num_extant:
+                    min_total_time = min(
+                        [i["total_time"] for i in current_lineages]
+                    )
+                    for remaining_lineage in current_lineages:
+                        parent = list(
+                            tree.predecessors(remaining_lineage["id"])
+                        )[0]
+                        tree.edges[parent, remaining_lineage["id"]][
+                            "weight"
+                        ] += (min_total_time - remaining_lineage["total_time"])
+                        tree.nodes[remaining_lineage["id"]][
+                            "birth_scale"
+                        ] = tree.nodes[parent]["birth_scale"]
+                    break
+            # If extant tips are the stopping criteria, pop the minimum age
+            # lineage at each step
+            if num_extant:
+                min_age_ind = np.argmin(
+                    [i["total_time"] for i in current_lineages]
+                )
+                lineage = current_lineages.pop(min_age_ind)
+            else:
+                lineage = current_lineages.pop(0)
+            for _ in range(2):
+                unique_id = sample_event(unique_id, lineage)
+
+        # Prune dead lineages and collapse resulting unifurcations
+        if death_waiting_dist and len(tree.nodes) > 1:
+            if experiment_time:
+                for i in list(tree.nodes):
+                    if (
+                        tree.out_degree(i) == 0
+                        and tree.nodes[i]["total_time"] < experiment_time
+                    ):
+                        self.remove_and_prune_lineage(i, tree)
+            if num_extant:
+                surviving_ids = [i["id"] for i in current_lineages]
+                for i in list(tree.nodes):
+                    if tree.out_degree(i) == 0 and i not in surviving_ids:
+                        self.remove_and_prune_lineage(i, tree)
+            if len(tree.nodes) > 1:
+                self.collapse_unifurcations(tree, source=1)
+
+        # If only implicit root remains after pruning dead lineages, error
+        if len(tree.nodes) == 1:
+            raise BirthDeathFitnessError(
+                "All lineages died before stopping condition"
+            )
+
+        return CassiopeiaTree(tree=tree)
+
+    def remove_and_prune_lineage(self, node: int, tree: nx.DiGraph):
```

Comment:
This method sounds more generally useful? Perhaps we can put it into `cassiopeia/data/utilities.py` or something?

Reply:
I agree. This should actually already be implemented in `cassiopeia.data.utilities`. 

At any rate, `ete3.Tree.prune` has very similar functionality. It'd be great to use one of these tools here.